bip-hugonguyen-bsms.mediawiki

```diff
@@ -0,0 +1,452 @@
+* A QR code.
+* Other formats.
+
+The flexibility in the data format allows each Signer to customize the User Experience based on its respective capabilities.
+
+====Key Derivation====
+The key derivation function is [https://tools.ietf.org/html/rfc2898 PBKDF2], with PRF = SHA512. Specifically:
+
+<tt>DKey = PBKDF2(PRF, Password, Salt, c, dkLen)</tt>
+
+Whereas:
+
+* PRF = SHA512
+* Password = "No SPOF"
+* Salt = <tt>TOKEN</tt>
+* c = 2048
+* dkLen = 256
+* DKey = Derived <tt>ENCRYPTION_KEY</tt>
+
+====Encryption Scheme====
+The encryption scheme is [https://tools.ietf.org/html/rfc3686 AES-256-CTR].
+
+<tt>MAC = HMAC-SHA256(HMAC_Key, hex-encoded TOKEN || Data)</tt>
+
+<tt>IV = First 16 bytes of MAC</tt>
+
+<tt>Ciphertext = AES-256-CTR-Encrypt(Plaintext, DKey, IV)</tt>
+
+<tt>Plaintext = AES-256-CTR-Decrypt(Ciphertext, DKey, IV)</tt>
+
+Whereas:
+* DKey = <tt>ENCRYPTION_KEY</tt>
+* HMAC_Key = SHA256(<tt>ENCRYPTION_KEY</tt>)
+* Data = the plaintext, e.g. the entire key record in round 1 and the entire descriptor record in round 2
+
+The <tt>MAC</tt> is to be sent along with the key and descriptor record, as specified above. Because it is a <tt>MAC</tt> over the entire plaintext, this is essentially an [https://en.wikipedia.org/wiki/Authenticated_encryption#Encrypt-and-MAC_(E&M) Encrypt-and-MAC] form of authenticated encryption.
+
+===Descriptor Template===
+The output descriptor language only supports one-dimensional lists. This proposal introduces a descriptor template to represent multi-dimensional lists:
+
+<tt>XPUB/**</tt>
+
+Whereas <tt>/**</tt> can be replaced by any number of derivation path restrictions.
+
+A descriptor template must be accompanied by derivation path restrictions. Signers should expand the template into concrete descriptors by replacing <tt>/**</tt> with the restrictions.
+
+For example, the following template and derivation path restrictions:
+* <tt>wsh(sortedmulti(2,XPUB1/**,XPUB2/**))</tt>
+* <tt>/0/*,/1/*</tt>
+
+Should translate to two concrete descriptors:
+* <tt>wsh(sortedmulti(2,XPUB1/0/*,XPUB2/0/*))</tt>
+* <tt>wsh(sortedmulti(2,XPUB1/1/*,XPUB2/1/*))</tt>
+
+==QR Codes==
+For signers that use QR codes to transmit data, key and descriptor records can be converted to QR codes, following [https://github.com/BlockchainCommons/Research/blob/master/papers/bcr-2020-005-ur.md the BCR standard].
+
+Also refer to [https://github.com/BlockchainCommons/Research/blob/master/papers/bcr-2020-015-account.md UR Type Definition for BIP44 Accounts] and [https://github.com/BlockchainCommons/Research/blob/master/papers/bcr-2020-010-output-desc.md UR Type Definition for Bitcoin Output Descriptors] for more details.
+
+==Compatibility==
```

Comment:
Can you add something about how this interacts with existing multisig technology?

Reply:
@luke-jr done https://github.com/bitcoin/bips/pull/1097/commits/83aa04776d38e18f03cd766102cfcefb5b244c9f

Reply:
I mean BSMS compatibility with current multisig hardware/softfork/firmware. It sounds like it's simply not compatible (since you are advising all users must upgrade before using BSMS)?

Reply:
@luke-jr BSMS is opt-in so it doesn't really affect existing multisig wallets (unlike many changes at the Bitcoin protocol layer where everyone has to upgrade). If you already had a multisig wallet, you can continue using it as-is, with the caveat that existing multisig solutions on the market are full of pitfalls. @benma wrote an article on this topic last year: https://benma.github.io/2020/11/05/multisig-xpubs-verification.html. We also discovered a related vulnerability that motivated us to write this standard in the first place. See https://medium.com/nunchuk/the-missing-multisig-standard-5b4092e3ea92.

To use BSMS, vendors (such as Coldcard, Trezor, BitBox02, Ledger, Cobo etc.) first have to update their firmware to add support for BSMS. Then the users can create new multisig wallets or upgrade their existing ones to use BSMS. But it's entirely opt-in.

Reply:
Yes, this kind of stuff should be in the BIP

Reply:
@luke-jr added some more details here https://github.com/bitcoin/bips/pull/1097/commits/f13cd8dde9e67003fcc952de02d0b63c4db5bbac. Note that I'm trying to use a broader language so as to avoid referring to any vendors, or types of vendors specifically.

Reply:
@luke-jr https://github.com/bitcoin/bips/pull/1097/commits/229a60bf8e180f6d78acce111018d261b8327dec